### PR TITLE
Linux & OS X fixes

### PR DESCRIPTION
**Pull request by almightykiwi**
Hi! I'm back.

Here are yet another bunch of fixes on Linux (and OS X!), mostly minor things, I think, except maybe a few methods in allocators.

I'm curious to know if it builds on Windows. (On OS X, StdAlloc::construct(U* p, Args&&... args) yields a build error.)

Best,
Marc

### CODE REVIEW

**Comment by almightykiwi** [comment]
Hmmm, have you changed the CLA thing? I'm told that I haven't signed it, but I don't think I did anything different from last time.

**Comment by MarcoROG** [comment]
Yes, CLA system has been changed, I suggest signing it again

**Comment by BearishSun** [comment]
Hey! Thanks for your contribution, once more. I did find a few issues, I'm leaving comments on the lines in question.

File: Source/BansheeCore/Source/BsProfilerCPU.cpp [hunk 1]
```diff
@@ -176,7 +176,7 @@ namespace BansheeEngine
 
 		activeBlock = ActiveBlock(ActiveSamplingType::Basic, rootBlock);
 		if (activeBlocks == nullptr)
-			activeBlocks = frameAlloc.alloc<Stack<ActiveBlock, StdFrameAlloc<ActiveBlock>>>(&frameAlloc);
+			activeBlocks = frameAlloc.alloc<Stack<ActiveBlock, StdFrameAlloc<ActiveBlock>>>();
```
Comment (BearishSun) [inline Source/BansheeCore/Source/BsProfilerCPU.cpp]: The lack of the &frameAlloc parameter is causing an access violation (it's trying to access a nullptr)

File: Source/BansheeUtility/Include/BsFrameAlloc.h [hunk 2]
```diff
@@ -151,6 +154,12 @@ namespace BansheeEngine
 	{
 	public:
 		typedef T value_type;
+		typedef value_type* pointer;
+		typedef const value_type* const_pointer;
+		typedef value_type& reference;
+		typedef const value_type& const_reference;
+		typedef std::size_t size_type;
+		typedef std::size_t difference_type;
```
Comment (BearishSun) [inline Source/BansheeUtility/Include/BsFrameAlloc.h]: Can you change the type of difference_type to std::ptrdiff_t. std::size_t is generating a lot of warnings in MSVC.

File: Source/BansheeCore/Include/BsGpuProgram.h [hunk 3]
```diff
@@ -231,4 +231,19 @@ namespace BansheeEngine
...
+namespace std
+{
+/** Hash value generator for GpuProgramProfile. */
+template<>
+struct hash<BansheeEngine::GpuProgramProfile>
+{
+	size_t operator()(const BansheeEngine::GpuProgramProfile& profile) const
+	{
+		size_t hash = 0;
+		BansheeEngine::hash_combine(hash, profile);
```
Comment (BearishSun) [inline Source/BansheeCore/Include/BsGpuProgram.h]: This is causing a stack overflow. Casting profile to integer should fix it.

File: Source/BansheeUtility/Include/BsFrameAlloc.h (hunk 2 above)
Comment (almightykiwi) [inline Source/BansheeUtility/Include/BsFrameAlloc.h]: Yep.

File: Source/BansheeCore/Include/BsGpuProgram.h (hunk 3 above)
Comment (almightykiwi) [inline Source/BansheeCore/Include/BsGpuProgram.h]: Okay.

File: Source/BansheeCore/Source/BsProfilerCPU.cpp (hunk 1 above)
Comment (almightykiwi) [inline Source/BansheeCore/Source/BsProfilerCPU.cpp]: Ok, I simply removed this commit for now.

**Comment by almightykiwi** [comment]
There, I addressed the issues you notice. If there is anything else don't hesitate.

**Comment by BearishSun** [comment]
Looking good now, tested on Windows.

If you are interested in actively working on Banshee's Linux and/or OS X ports let me know and you can join us on chat.

**Pull request merged**